File: apps/wallet-mobile/etrid-wallet/services/DAppBrowserService.ts
/**
 * DApp Browser Service
 * Handles Web3 provider injection and dApp interactions
 */

import {
  DAppRequest,
  ConnectedDApp,
  DAppPermission,
  TransactionRequest,
  TransactionSimulation,
  Web3Provider,
} from '@/types/dapp';

export class DAppBrowserService {
  private connectedDApps: Map<string, ConnectedDApp> = new Map();
  private pendingRequests: Map<string, DAppRequest> = new Map();

  /**
   * Inject Web3 provider into WebView
   */
  async injectWeb3Provider(webView: any): Promise<void> {
    const providerScript = `
      (function() {
        const etridProvider = {
          isEtrid: true,
          isMetaMask: true, // MetaMask compatibility
          chainId: null,
          selectedAddress: null,

          request: async function(args) {
            return new Promise((resolve, reject) => {
              const requestId = Date.now().toString();
              const request = {
                id: requestId,
                method: args.method,
                params: args.params || []
              };

              // Post message to native app
              window.ReactNativeWebView.postMessage(JSON.stringify({
                type: 'web3_request',
                data: request
              }));

              // Store resolver
              window.__etridRequests = window.__etridRequests || {};
              window.__etridRequests[requestId] = { resolve, reject };
            });
          },

          on: function(event, handler) {
            window.addEventListener('etrid_' + event, handler);
          },

          removeListener: function(event, handler) {
            window.removeEventListener('etrid_' + event, handler);
          }
        };

        window.ethereum = etridProvider;
        window.web3 = { currentProvider: etridProvider };

        // Response handler
        window.addEventListener('message', function(event) {
          try {
            const message = JSON.parse(event.data);
            if (message.type === 'web3_response') {
              const { requestId, result, error } = message.data;
              const request = window.__etridRequests[requestId];
              if (request) {
                if (error) {
                  request.reject(new Error(error));
                } else {
                  request.resolve(result);
                }
                delete window.__etridRequests[requestId];
              }
            }
          } catch (e) {}
        });

        // Emit connected event
        window.dispatchEvent(new Event('ethereum#initialized'));
      })();
    `;

    // Inject script into WebView
    if (webView && webView.injectJavaScript) {
      webView.injectJavaScript(providerScript);
    }
  }

  /**
   * Handle dApp request from WebView
   */
  async handleDAppRequest(request: DAppRequest): Promise<any> {
    const { method, params } = request;

    switch (method) {
      case 'eth_requestAccounts':
        return this.handleRequestAccounts(request);

      case 'eth_accounts':
        return this.handleGetAccounts(request);

      case 'eth_chainId':
        return this.handleGetChainId();

      case 'eth_getBalance':
        return this.handleGetBalance(params[0]);

      case 'eth_sendTransaction':
        return this.handleSendTransaction(request, params[0]);

      case 'eth_signTransaction':
        return this.handleSignTransaction(request, params[0]);

      case 'eth_sign':
        return this.handleSign(request, params[0], params[1]);

      case 'personal_sign':
        return this.handlePersonalSign(request, params[0], params[1]);

      case 'eth_signTypedData':
      case 'eth_signTypedData_v4':
        return this.handleSignTypedData(request, params[0], params[1]);

      default:
        throw new Error(`Unsupported method: ${method}`);
    }
  }

  /**
   * Request account access
   */
  private async handleRequestAccounts(request: DAppRequest): Promise<string[]> {
    // Check if already connected
    const connected = this.connectedDApps.get(request.dAppUrl);
    if (connected) {
      return [connected.permissions[0]?.type || ''];
    }

    // Request user approval (this would show a modal in the UI)
    const approved = await this.requestUserApproval(request.dAppUrl, [
      { type: 'access_address', grantedAt: new Date() },
    ]);

    if (!approved) {
      throw new Error('User rejected request');
    }

    // Return connected account
    const account = await this.getCurrentAccount();
    return [account];
  }

  /**
   * Get connected accounts
   */
  private async handleGetAccounts(request: DAppRequest): Promise<string[]> {
    const connected = this.connectedDApps.get(request.dAppUrl);
    if (!connected) {
      return [];
    }
    const account = await this.getCurrentAccount();
    return [account];
  }

  /**
   * Get chain ID
   */
  private async handleGetChainId(): Promise<string> {
    // Return Etrid chain ID (placeholder)
    return '0x1'; // Ethereum mainnet format
  }

  /**
   * Get balance
   */
  private async handleGetBalance(address: string): Promise<string> {
    // TODO: Implement actual balance fetching
    // This would integrate with your wallet's balance service
    return '0x0';
  }

  /**
   * Send transaction
   */
  private async handleSendTransaction(
    request: DAppRequest,
    transaction: TransactionRequest
  ): Promise<string> {
    // Check permissions
    const hasPermission = await this.checkPermission(
      request.dAppUrl,
      'sign_transaction'
    );
    if (!hasPermission) {
      throw new Error('Permission denied');
    }

    // Simulate transaction to detect issues
    const simulation = await this.simulateTransaction(transaction);
    if (!simulation.success) {
      throw new Error('Transaction simulation failed');
    }

    // Request user approval with simulation results
    const approved = await this.requestTransactionApproval(
      request.dAppUrl,
      transaction,
      simulation
    );

    if (!approved) {
      throw new Error('User rejected transaction');
    }

    // Send transaction (this would integrate with your wallet's transaction service)
    const txHash = await this.sendTransaction(transaction);
    return txHash;
  }

  /**
   * Sign transaction (without sending)
   */
  private async handleSignTransaction(
    request: DAppRequest,
    transaction: TransactionRequest
  ): Promise<string> {
    // Similar to sendTransaction but only returns signed transaction
    const hasPermission = await this.checkPermission(
      request.dAppUrl,
      'sign_transaction'
    );
    if (!hasPermission) {
      throw new Error('Permission denied');
    }

    const approved = await this.requestTransactionApproval(
      request.dAppUrl,
      transaction
    );

    if (!approved) {
      throw new Error('User rejected transaction');
    }

    // Sign transaction
    const signedTx = await this.signTransaction(transaction);
    return signedTx;
  }

  /**
   * Sign message (eth_sign)
   */
  private async handleSign(
    request: DAppRequest,
    address: string,
    message: string
  ): Promise<string> {
    const hasPermission = await this.checkPermission(
      request.dAppUrl,
      'sign_transaction'
    );
    if (!hasPermission) {
      throw new Error('Permission denied');
    }

    const approved = await this.requestSignatureApproval(
      request.dAppUrl,
      message,
      'message'
    );

    if (!approved) {
      throw new Error('User rejected signature');
    }

    const signature = await this.signMessage(message);
    return signature;
  }

  /**
   * Sign message (personal_sign)
   */
  private async handlePersonalSign(
    request: DAppRequest,
    message: string,
    address: string
  ): Promise<string> {
    return this.handleSign(request, address, message);
  }

  /**
   * Sign typed data (EIP-712)
   */
  private async handleSignTypedData(
    request: DAppRequest,
    address: string,
    typedData: string
  ): Promise<string> {
    const hasPermission = await this.checkPermission(
      request.dAppUrl,
      'sign_transaction'
    );
    if (!hasPermission) {
      throw new Error('Permission denied');
    }

    const approved = await this.requestSignatureApproval(
      request.dAppUrl,
      typedData,
      'typed_data'
    );

    if (!approved) {
      throw new Error('User rejected signature');
    }

    const signature = await this.signTypedData(typedData);
    return signature;
  }

  /**
   * Get connected dApps
   */
  async getConnectedDApps(): Promise<ConnectedDApp[]> {
    return Array.from(this.connectedDApps.values());
  }

  /**
   * Disconnect dApp
   */
  async disconnectDApp(dAppUrl: string): Promise<void> {
    this.connectedDApps.delete(dAppUrl);
    // TODO: Persist to storage
  }

  /**
   * Check if dApp has permission
   */
  private async checkPermission(
    dAppUrl: string,
    permissionType: string
  ): Promise<boolean> {
    const dApp = this.connectedDApps.get(dAppUrl);
    if (!dApp) return false;
    return dApp.permissions.some((p) => p.type === permissionType);
  }

  /**
   * Simulate transaction to detect issues
   */
  private async simulateTransaction(
    transaction: TransactionRequest
  ): Promise<TransactionSimulation> {
    // TODO: Implement actual transaction simulation
    // This would call a simulation service to preview the transaction outcome
    return {
      success: true,
      gasEstimate: '21000',
      balanceChanges: [],
      warnings: [],
    };
  }

  // Mock methods that would be implemented with actual wallet integration
  private async getCurrentAccount(): Promise<string> {
    // TODO: Get from wallet service
    return '0x0000000000000000000000000000000000000000';
  }

  private async requestUserApproval(
    dAppUrl: string,
    permissions: DAppPermission[]
  ): Promise<boolean> {
    // TODO: Show approval modal in UI
    return true;
  }

  private async requestTransactionApproval(
    dAppUrl: string,
    transaction: TransactionRequest,
    simulation?: TransactionSimulation
  ): Promise<boolean> {
    // TODO: Show transaction approval modal in UI
    return true;
  }

  private async requestSignatureApproval(
    dAppUrl: string,
    data: string,
    type: string
  ): Promise<boolean> {
    // TODO: Show signature approval modal in UI
    return true;
  }

  private async sendTransaction(transaction: TransactionRequest): Promise<string> {
    // TODO: Implement actual transaction sending
    return '0x0000000000000000000000000000000000000000000000000000000000000000';
  }

  private async signTransaction(transaction: TransactionRequest): Promise<string> {
    // TODO: Implement actual transaction signing
    return '0x0000000000000000000000000000000000000000000000000000000000000000';
  }

  private async signMessage(message: string): Promise<string> {
    // TODO: Implement actual message signing
    return '0x0000000000000000000000000000000000000000000000000000000000000000';
  }

  private async signTypedData(typedData: string): Promise<string> {
    // TODO: Implement actual typed data signing
    return '0x0000000000000000000000000000000000000000000000000000000000000000';
  }
}

export const dAppBrowserService = new DAppBrowserService();

File: apps/EtridWalletApp/EtridWallet/Services/DAppBrowserService.swift
//
//  DAppBrowserService.swift
//  EtridWallet
//
//  Service for dApp browser and Web3 provider injection
//

import Foundation
import WebKit
import Combine

@MainActor
class DAppBrowserService: ObservableObject {
    static let shared = DAppBrowserService()

    @Published var connectedDApps: [ConnectedDApp] = []
    @Published var bookmarks: [Bookmark] = []
    @Published var pendingRequest: Web3Request?
    @Published var isProcessingRequest = false

    private var requestHandlers: [String: (Result<String, Error>) -> Void] = [:]
    private let storageKey = "connected_dapps"
    private let bookmarksKey = "dapp_bookmarks"

    private init() {
        loadConnectedDApps()
        loadBookmarks()
    }

    // MARK: - Web3 Provider Injection

    func injectWeb3Provider(into webView: WKWebView) async throws {
        let providerScript = """
        (function() {
            const etridProvider = {
                isEtrid: true,
                isMetaMask: true,
                chainId: null,
                selectedAddress: null,
                networkVersion: null,

                request: async function(args) {
                    return new Promise((resolve, reject) => {
                        const requestId = Date.now().toString() + Math.random().toString();
                        const request = {
                            id: requestId,
                            method: args.method,
                            params: args.params || []
                        };

                        window.webkit.messageHandlers.web3.postMessage({
                            type: 'web3_request',
                            data: request
                        });

                        window.__etridRequests = window.__etridRequests || {};
                        window.__etridRequests[requestId] = { resolve, reject };

                        setTimeout(() => {
                            if (window.__etridRequests[requestId]) {
                                reject(new Error('Request timeout'));
                                delete window.__etridRequests[requestId];
                            }
                        }, 60000);
                    });
                },

                sendAsync: function(payload, callback) {
                    this.request(payload)
                        .then(result => callback(null, { id: payload.id, jsonrpc: '2.0', result }))
                        .catch(error => callback(error, null));
                },

                send: function(methodOrPayload, paramsOrCallback) {
                    if (typeof methodOrPayload === 'string') {
                        return this.request({ method: methodOrPayload, params: paramsOrCallback });
                    }
                    return this.sendAsync(methodOrPayload, paramsOrCallback);
                },

                on: function(event, handler) {
                    window.addEventListener('etrid_' + event, handler);
                },

                removeListener: function(event, handler) {
                    window.removeEventListener('etrid_' + event, handler);
                },

                isConnected: function() {
                    return true;
                }
            };

            window.ethereum = etridProvider;
            window.web3 = { currentProvider: etridProvider };

            window.addEventListener('message', function(event) {
                try {
                    const message = event.data;
                    if (message && message.type === 'web3_response') {
                        const { requestId, result, error } = message.data;
                        const request = window.__etridRequests[requestId];
                        if (request) {
                            if (error) {
                                request.reject(new Error(error));
                            } else {
                                request.resolve(result);
                            }
                            delete window.__etridRequests[requestId];
                        }
                    }
                } catch (e) {
                    console.error('Web3 response error:', e);
                }
            });

            window.dispatchEvent(new Event('ethereum#initialized'));
        })();
        """

        try await webView.evaluateJavaScript(providerScript)
    }

    // MARK: - Request Handling

    func handleWeb3Request(_ request: Web3Request) async throws -> String {
        switch request.method {
        case "eth_requestAccounts":
            return try await handleRequestAccounts(request)
        case "eth_accounts":
            return try await handleGetAccounts(request)
        case "eth_chainId":
            return try await handleGetChainId()
        case "eth_sendTransaction":
            return try await handleSendTransaction(request)
        case "personal_sign":
            return try await handlePersonalSign(request)
        case "eth_signTypedData", "eth_signTypedData_v4":
            return try await handleSignTypedData(request)
        default:
            throw WalletError.unknown("Unsupported method: \(request.method)")
        }
    }

    private func handleRequestAccounts(_ request: Web3Request) async throws -> String {
        // Check if already connected
        if let connected = connectedDApps.first(where: { $0.dAppUrl == request.dAppUrl }) {
            let address = try await getCurrentAddress()
            return "[\"\(address)\"]"
        }

        // Request user approval
        pendingRequest = request
        let approved = try await waitForUserApproval()

        guard approved else {
            throw WalletError.unknown("User rejected")
        }

        // Add to connected dApps
        let connection = ConnectedDApp(
            id: UUID().uuidString,
            dAppUrl: request.dAppUrl,
            dAppName: request.dAppName,
            dAppIcon: nil,
            permissions: [DAppPermission(type: .accessAddress, grantedAt: Date())],
            connectedAt: Date(),
            lastUsed: Date()
        )
        connectedDApps.append(connection)
        saveConnectedDApps()

        let address = try await getCurrentAddress()
        return "[\"\(address)\"]"
    }

    private func handleGetAccounts(_ request: Web3Request) async throws -> String {
        if connectedDApps.contains(where: { $0.dAppUrl == request.dAppUrl }) {
            let address = try await getCurrentAddress()
            return "[\"\(address)\"]"
        }
        return "[]"
    }

    private func handleGetChainId() async throws -> String {
        // Return Etrid chain ID or current network
        return "\"0x1\"" // Ethereum mainnet format
    }

    private func handleSendTransaction(_ request: Web3Request) async throws -> String {
        pendingRequest = request
        let approved = try await waitForUserApproval()

        guard approved else {
            throw WalletError.unknown("User rejected")
        }

        // TODO: Integrate with TransactionManager
        let txHash = "0x" + UUID().uuidString.replacingOccurrences(of: "-", with: "")
        return "\"\(txHash)\""
    }

    private func handlePersonalSign(_ request: Web3Request) async throws -> String {
        pendingRequest = request
        let approved = try await waitForUserApproval()

        guard approved else {
            throw WalletError.unknown("User rejected")
        }

        // TODO: Integrate with CryptoManager for signing
        let signature = "0x" + String(repeating: "0", count: 130)
        return "\"\(signature)\""
    }

    private func handleSignTypedData(_ request: Web3Request) async throws -> String {
        pendingRequest = request
        let approved = try await waitForUserApproval()

        guard approved else {
            throw WalletError.unknown("User rejected")
        }

        // TODO: Implement EIP-712 signing
        let signature = "0x" + String(repeating: "0", count: 130)
        return "\"\(signature)\""
    }

    // MARK: - Response Handling

    func sendResponse(_ response: Web3Response, to webView: WKWebView) {
        let responseJSON: String
        if let result = response.result {
            responseJSON = "{\"type\":\"web3_response\",\"data\":{\"requestId\":\"\(response.requestId)\",\"result\":\(result)}}"
        } else if let error = response.error {
            responseJSON = "{\"type\":\"web3_response\",\"data\":{\"requestId\":\"\(response.requestId)\",\"error\":\"\(error)\"}}"
        } else {
            return
        }

        let script = "window.postMessage(\(responseJSON), '*');"
        webView.evaluateJavaScript(script) { _, error in
            if let error = error {
                print("Error sending response: \(error)")
            }
        }
    }

    // MARK: - Bookmarks

    func addBookmark(name: String, url: String, iconUrl: String?) {
        let bookmark = Bookmark(
            id: UUID().uuidString,
            name: name,
            url: url,
            iconUrl: iconUrl,
            createdAt: Date()
        )
        bookmarks.append(bookmark)
        saveBookmarks()
    }

    func removeBookmark(_ bookmark: Bookmark) {
        bookmarks.removeAll { $0.id == bookmark.id }
        saveBookmarks()
    }

    // MARK: - Connected DApps

    func disconnectDApp(_ dapp: ConnectedDApp) {
        connectedDApps.removeAll { $0.id == dapp.id }
        saveConnectedDApps()
    }

    // MARK: - Phishing Detection

    func checkPhishing(url: String) -> PhishingCheck {
        // Basic phishing detection
        let suspiciousPatterns = [
            "metamask-",
            "uniswap-",
            "opensea-",
            "coinbase-",
            ".tk",
            ".ml",
            ".ga"
        ]

        var warnings: [PhishingCheck.SecurityWarning] = []
        var risk = PhishingCheck.RiskLevel.safe

        for pattern in suspiciousPatterns {
            if url.lowercased().contains(pattern) {
                warnings.append(PhishingCheck.SecurityWarning(
                    title: "Suspicious Domain",
                    description: "This domain contains suspicious patterns",
                    severity: .warning
                ))
                risk = .medium
                break
            }
        }

        if !url.hasPrefix("https://") {
            warnings.append(PhishingCheck.SecurityWarning(
                title: "Insecure Connection",
                description: "This site does not use HTTPS",
                severity: .critical
            ))
            risk = .high
        }

        return PhishingCheck(
            url: url,
            isSafe: risk == .safe,
            warnings: warnings,
            risk: risk
        )
    }

    // MARK: - Helpers

    private func getCurrentAddress() async throws -> String {
        // TODO: Get from WalletManager
        return "0x742d35Cc6634C0532925a3b844Bc454e4438f44e"
    }

    private func waitForUserApproval() async throws -> Bool {
        // Wait for user interaction (approved/rejected)
        // This would be triggered by UI interaction
        return true
    }

    // MARK: - Storage

    private func saveConnectedDApps() {
        if let encoded = try? JSONEncoder().encode(connectedDApps) {
            UserDefaults.standard.set(encoded, forKey: storageKey)
        }
    }

    private func loadConnectedDApps() {
        if let data = UserDefaults.standard.data(forKey: storageKey),
           let decoded = try? JSONDecoder().decode([ConnectedDApp].self, from: data) {
            connectedDApps = decoded
        }
    }

    private func saveBookmarks() {
        if let encoded = try? JSONEncoder().encode(bookmarks) {
            UserDefaults.standard.set(encoded, forKey: bookmarksKey)
        }
    }

    private func loadBookmarks() {
        if let data = UserDefaults.standard.data(forKey: bookmarksKey),
           let decoded = try? JSONDecoder().decode([Bookmark].self, from: data) {
            bookmarks = decoded
        }
    }
}
